drivers/spi/spi_nrfx_spim.c

```diff
@@ -480,140 +507,128 @@ static int spim_nrfx_pm_action(const struct device *dev,
+									       \
+	static void irq_connect##inst(void)				       \
 	{								       \
-		struct spi_nrfx_data *dev_data = dev->data;		       \
-		int err;						       \
-		IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_SPIM##idx),		       \
-			    DT_IRQ(SPIM(idx), priority),		       \
-			    nrfx_isr, nrfx_spim_##idx##_irq_handler, 0);       \
-		IF_ENABLED(CONFIG_PINCTRL, (				       \
-			const struct spi_nrfx_config *dev_config = dev->config;\
-			err = pinctrl_apply_state(dev_config->pcfg,	       \
-						  PINCTRL_STATE_DEFAULT);      \
-			if (err < 0) {					       \
-				return err;				       \
-			}						       \
-		))							       \
-		err = spi_context_cs_configure_all(&dev_data->ctx);	       \
-		if (err < 0) {						       \
-			return err;					       \
-		}							       \
-		spi_context_unlock_unconditionally(&dev_data->ctx);	       \
-		COND_CODE_1(CONFIG_SOC_NRF52832_ALLOW_SPIM_DESPITE_PAN_58,     \
-		(return anomaly_58_workaround_init(dev);),		       \
-		(return 0;))						       \
+		IRQ_CONNECT(DT_INST_IRQN(inst), DT_INST_IRQ(inst, priority),   \
+			    nrfx_isr,					       \
+			    SPI_IRQ_HANDLER(DT_INST_PROP(inst, periph_idx)),   \
+			    0);						       \
 	}								       \
+									       \
+	PM_DEVICE_DT_INST_DEFINE(inst, spim_nrfx_pm_action);		       \
 	IF_ENABLED(SPI_BUFFER_IN_RAM,					       \
-		(static uint8_t spim_##idx##_buffer			       \
+		   (static uint8_t spim_##inst##_buffer			       \
 			[CONFIG_SPI_NRFX_RAM_BUFFER_SIZE]		       \
-			SPIM_MEMORY_SECTION(idx);))			       \
-	static struct spi_nrfx_data spi_##idx##_data = {		       \
-		SPI_CONTEXT_INIT_LOCK(spi_##idx##_data, ctx),		       \
-		SPI_CONTEXT_INIT_SYNC(spi_##idx##_data, ctx),		       \
-		SPI_CONTEXT_CS_GPIOS_INITIALIZE(SPIM(idx), ctx)		       \
+			SPIM_MEMORY_SECTION(inst);))			       \
+	IF_ENABLED(CONFIG_PINCTRL, (PINCTRL_DT_INST_DEFINE(inst)));	       \
+									       \
+	static struct spi_nrfx_data data##inst = {			       \
+		SPI_CONTEXT_INIT_LOCK(data##inst, ctx),			       \
+		SPI_CONTEXT_INIT_SYNC(data##inst, ctx),			       \
+		SPI_CONTEXT_CS_GPIOS_INITIALIZE(DT_DRV_INST(inst), ctx)	       \
 		IF_ENABLED(SPI_BUFFER_IN_RAM,				       \
-			(.buffer = spim_##idx##_buffer,))		       \
-		.dev  = DEVICE_DT_GET(SPIM(idx)),			       \
+			   (.buffer = spim_##inst##_buffer,))		       \
+		.dev  = DEVICE_DT_INST_GET(inst),			       \
 		.busy = false,						       \
 	};								       \
-	IF_ENABLED(CONFIG_PINCTRL, (PINCTRL_DT_DEFINE(SPIM(idx))));	       \
-	static const struct spi_nrfx_config spi_##idx##z_config = {	       \
-		.spim = NRFX_SPIM_INSTANCE(idx),			       \
-		.max_chunk_len = (1 << SPIM##idx##_EASYDMA_MAXCNT_SIZE) - 1,   \
-		.max_freq = SPIM##idx##_MAX_DATARATE * 1000000,		       \
+									       \
+	static const struct spi_nrfx_config config##inst = {		       \
+		.spim = {						       \
+			.p_reg = (NRF_SPIM_Type *)DT_INST_REG_ADDR(inst),      \
+			.drv_inst_idx = DT_INST_PROP(inst, periph_idx),	       \
```

Comment:
This is incorrect. The index is not a peripheral instance index, it is an nrfx_spim driver instance index (e.g. it will be 0, when only SPIM3 is enabled to be handled by the driver).

Reply:
good point, however, `NRFX_SPIM_INSTANCE` doesn't take device address from DT, which needs to be fixed. Overall, I think that in this case, HAL is making drivers more complicated because it overlaps with functionality covered by Zephyr. This is in fact a common pattern with HALs like nrfx (stateful), they tend to be inefficient and induce bad design patterns. We also have some other notable inefficiencies like this: https://github.com/zephyrproject-rtos/zephyr/blob/88ca3aca98c0788e2c45db5c3fb9c8523293a8c0/drivers/i2c/i2c_nrfx_twim.c#L295-L300,ie what could have been a simple register write replaced by a full deinit/init sequence because HAL doesn't support updating frequency.